numba/tests/test_slices.py

```diff
@@ -60,27 +66,71 @@ def check(a, b, c, d, e, f):
+        a = np.arange(10)
         cfunc = jit(nopython=True)(slice_constructor)
-        for args, expected in [((), (0, maxposint, 1)),
-                               ((None, None), (0, maxposint, 1)),
-                               ((1, None), (1, maxposint, 1)),
-                               ((None, 2), (0, 2, 1)),
-                               ((1, 2), (1, 2, 1)),
-                               ((None, None, 3), (0, maxposint, 3)),
-                               ((None, 2, 3), (0, 2, 3)),
-                               ((1, None, 3), (1, maxposint, 3)),
-                               ((1, 2, 3), (1, 2, 3)),
-                               ((None, None, -1), (maxposint, maxnegint, -1)),
-                               ((10, None, -1), (10, maxnegint, -1)),
-                               ((None, 5, -1), (maxposint, 5, -1)),
-                               ((10, 5, -1), (10, 5, -1)),
-                               ]:
+        cfunc_use = jit(nopython=True)(slice_construct_and_use)
+        for args, expected in [
+            ((None,), (0, maxposint, 1)),
+            ((5,), (0, 5, 1)),
+            ((None, None), (0, maxposint, 1)),
+            ((1, None), (1, maxposint, 1)),
+            ((None, 2), (0, 2, 1)),
+            ((1, 2), (1, 2, 1)),
+            ((None, None, 3), (0, maxposint, 3)),
+            ((None, 2, 3), (0, 2, 3)),
+            ((1, None, 3), (1, maxposint, 3)),
+            ((1, 2, 3), (1, 2, 3)),
+            ((None, None, -1), (maxposint, maxnegint, -1)),
+            ((10, None, -1), (10, maxnegint, -1)),
+            ((None, 5, -1), (maxposint, 5, -1)),
+            ((10, 5, -1), (10, 5, -1)),
+        ]:
             got = cfunc(*args)
             self.assertPreciseEqual(got, expected)
+            usage = slice_construct_and_use(args, a)
+            cusage = cfunc_use(args, a)
+            self.assertPreciseEqual(usage, cusage)
+
+    def test_slice_constructor_cases(self):
+        """
+        Test that slice constructor behaves same in python and compiled code.
+        """
+        options = (None, -1, 0, 1)
+        arg_cases = chain.from_iterable(
+            product(options, repeat=n) for n in range(5)
+        )
+        array = np.arange(10)
+
+        cfunc = jit(nopython=True)(slice_construct_and_use)
+
+        self.disable_leak_check()
+        for args in arg_cases:
+            try:
+                expected = slice_construct_and_use(args, array)
+            except TypeError as py_type_e:
+                # Catch cases of 0, or more than 3 arguments.
+                # This becomes a typing error in numba
+                n_args = len(args)
+                self.assertRegexpMatches(
+                    str(py_type_e),
+                    r"slice expected at (most|least) (3|1) arguments, got {}"
+                    .format(n_args)
+                )
+                with self.assertRaises(TypingError) as numba_e:
+                    cfunc(args, array)
```

Comment:
You'll probably want to do an `assertIn` here too.

Reply:
I've added a check, but I don't want the test to fail if the formatting of `TypingError` changes. Is there a way to mock errors?

Reply:
Numba tests typically check error message strings, `TypingError` is really common and checking the string ensures it's been raised from the expected location.  A large number of examples exist in the unit test, but here's one to save looking  :) 
https://github.com/numba/numba/blob/24e5ee8444705707f55e6d6b65779e4d333be0d5/numba/tests/test_np_functions.py#L474-L479

Reply:
So this is weird. I did add checks for the contents of the error message in my last commit. But the link at the top of this thread is specific to an earlier commit, so that doesn't show up. IIRC, this normally isn't the case.

Anyways: here's a link to the changes where the checks are visible: https://github.com/numba/numba/pull/4779/commits/24e5ee8444705707f55e6d6b65779e4d333be0d5

On another topic, in the future would it be possible to throw a different error from the typing stage? For example, if I wanted to throw the same error as python does.

Reply:
You can raise any error you like, it is mostly convention to raise a `TypingError` during typing stage.